File: driftwatch/backend/internal/api/github.go
package api

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/souvik-biswas-dev/driftwatch/internal/crypto"
)

// handleListUserRepos returns the authenticated user's GitHub repos so the
// dashboard can show a picker instead of free-text inputs.
func (a *API) handleListUserRepos(c *gin.Context) {
	userID, ok := currentUserID(c)
	if !ok {
		respondError(c, http.StatusUnauthorized, "auth required", "AUTH_MISSING")
		return
	}

	enc, err := a.queries.GetUserGithubToken(c.Request.Context(), userID)
	if err != nil || enc == "" {
		respondError(c, http.StatusBadRequest, "no GitHub token — log in with GitHub first", "NO_GITHUB_TOKEN")
		return
	}
	token, err := crypto.Decrypt(enc)
	if err != nil {
		respondError(c, http.StatusInternalServerError, "could not read GitHub token", "DECRYPT_ERROR")
		return
	}

	repos, err := a.githubClient.ListUserRepos(c.Request.Context(), token)
	if err != nil {
		respondError(c, http.StatusBadGateway, "could not list GitHub repos", "GITHUB_ERROR")
		return
	}
	respond(c, http.StatusOK, repos, "")
}

// handleListRepoBranches returns the branches for a specific repo.
func (a *API) handleListRepoBranches(c *gin.Context) {
	userID, ok := currentUserID(c)
	if !ok {
		respondError(c, http.StatusUnauthorized, "auth required", "AUTH_MISSING")
		return
	}

	owner := c.Param("owner")
	repo := c.Param("repo")

	enc, err := a.queries.GetUserGithubToken(c.Request.Context(), userID)
	if err != nil || enc == "" {
		respondError(c, http.StatusBadRequest, "no GitHub token — log in with GitHub first", "NO_GITHUB_TOKEN")
		return
	}
	token, err := crypto.Decrypt(enc)
	if err != nil {
		respondError(c, http.StatusInternalServerError, "could not read GitHub token", "DECRYPT_ERROR")
		return
	}

	branches, err := a.githubClient.ListRepoBranches(c.Request.Context(), owner, repo, token)
	if err != nil {
		respondError(c, http.StatusBadGateway, "could not list branches", "GITHUB_ERROR")
		return
	}
	respond(c, http.StatusOK, branches, "")
}

File: driftwatch/backend/src/api/routes/github.ts
import { Router } from 'express';

import { getUserGithubToken } from '../../db/users.js';
import { decrypt } from '../../services/crypto.js';
import { currentUserID } from '../auth.js';
import type { ApiDeps } from '../deps.js';
import { respond, respondError } from '../respond.js';

/**
 * Repo/branch pickers for the "new project" flow, read with the user's own
 * OAuth token so private repos show up.
 */
export function githubRoutes(deps: ApiDeps): Router {
	const r = Router();

	r.get('/github/repos', async (req, res) => {
		const userID = currentUserID(req, res);
		if (!userID) return;

		const token = await resolveUserToken(deps, userID, res);
		if (token === null) return;

		try {
			respond(res, 200, await deps.github.listUserRepos(token));
		} catch {
			respondError(res, 502, 'could not list GitHub repos', 'GITHUB_ERROR');
		}
	});

	r.get('/github/repos/:owner/:repo/branches', async (req, res) => {
		const userID = currentUserID(req, res);
		if (!userID) return;

		const token = await resolveUserToken(deps, userID, res);
		if (token === null) return;

		try {
			const branches = await deps.github.listRepoBranches(
				req.params.owner,
				req.params.repo,
				token
			);
			respond(res, 200, branches);
		} catch {
			respondError(res, 502, 'could not list branches', 'GITHUB_ERROR');
		}
	});

	return r;
}

/**
 * Loads and decrypts the user's stored OAuth token. Writes the error response
 * and returns null when it isn't usable.
 */
async function resolveUserToken(
	deps: ApiDeps,
	userID: string,
	res: Parameters<typeof respondError>[0]
): Promise<string | null> {
	let enc = '';
	try {
		enc = await getUserGithubToken(deps.db, userID);
	} catch {
		enc = '';
	}
	if (!enc) {
		respondError(
			res,
			400,
			'no GitHub token — log in with GitHub first',
			'NO_GITHUB_TOKEN'
		);
		return null;
	}

	try {
		return decrypt(enc);
	} catch {
		respondError(res, 500, 'could not read GitHub token', 'DECRYPT_ERROR');
		return null;
	}
}
